[5주차] Makes Restaurants Lookup App

TaeWoongKim2 opened the pull request:
## 과제 목표

- [x] TDD로 기능 구현
- [x] 레스토랑 카테고리 목록 보기 기능
- [x] 레스토랑 목록 보기 기능
- [x] 레스토랑 지역 목록 보기 기능
- [x] 커버리지 100% 달성

## 제한 조건

**모든 기능은 TDD로 구현**해야 하며, **Component와 Reducer에 대한 테스트**가 존재해야 한다. 
먼저 **테스트를 추가하고 commit 하고, 그 후 실제 기능을 추가하여 commit** 하도록 한다.

commit 메시지 로그는 아래와 같이 남기자:

> Restaurant에 대한 테스트를 추가한다.
Restaurant을 구현한다.
Restaurant를 리팩토링한다.
...

## 진행 사항

Make a test for App component.
App has a title named 'Restaurants'.

TaeWoongKim2 committed:
[Red] App에 대한 테스트 작성

Make a test for App component.
App has a title named 'Restaurants'.

TaeWoongKim2 committed:
RestaurantRegionContainer에 대한 테스트 작성

레스토랑 지역 정보를 나타낼 컨테이너에 대한 테스트 작성.

TaeWoongKim2 committed:
[Green] RestaurantRegionContainer 컴포넌트 구현

작성된 RestaurantRegionContainer 테스트에 따라 컴포넌트 구현.

TaeWoongKim2 committed:
Regions에 대한 테스트 작성

음식점 지역 정보를 리스터업할 Regions 컴포넌트에 대한 테스트 작성.
1st, with one or more regions.
2nd, without region.

TaeWoongKim2 committed:
[Green] Regions 컴포넌트 생성

레스토랑 지역 목록을 보여주는 컴포넌트 생성.
Regions 테스트를 기반으로 생성.
*TODO: (V)에 대한 테스트 필요!

TaeWoongKim2 committed:
[Refactor] Regions에 대한 테스트 추가

버튼 클릭시 (V) 체크 표시 테스트 케이스 추가.

TaeWoongKim2 committed:
[Refector] App & RestaurantRegionContainer 리팩토링

1st, Regions component into RestaurantRegionContainer.
2nd, RestaurantRegionContainer component into App.
3nd, Refactoring test case about RestaurantRegionContainer.

TaeWoongKim2 committed:
[Red] Reducer에 대한 테스트 작성

1st, Make store.js
2nd, Write test for reducer
3nd, Make reducer.js

TaeWoongKim2 committed:
[Green] reducer initial state 구현

initialState만 구현.
*TODO: selectRegion에 대한 테스트 및 구현!

TaeWoongKim2 committed:
[Refactor] actions.js 생성

Action creators를 모아둘 actions.js 생성.

TaeWoongKim2 committed:
[Green] action & recuder 리팩토링

1nd, initalState 변경으로 인한 리팩토링.
2nd, selectRegion에 대한 테스트 구현.

TaeWoongKim2 committed:
[Green] reducer & action 구현.

selectRegion에 대한 리듀서 및 액션 구현.

TaeWoongKim2 committed:
[Red] RestaurantRegionContainer에 대한 selectRegion 테스트 추가

If selects a region, check the region.
*TODO: Refector RestaurantRegionContainer and Regions, App.

TaeWoongKim2 committed:
[Green] RestaurantRegionContainer 컴포넌트에 리덕스 적용

useDispatch, useSelector를 통해 selectRegion action creator를 dispatch에 등록.

TaeWoongKim2 committed:
[Refactor] 리덕스 적용에 따른 리팩토링

1st, RestaurantRegionContainer 컴포넌트 리팩토링.
2nd, Regions에 대한 테스트 리팩토링.
3rd, Regions 컴포넌트 리팩토링.

TaeWoongKim2 committed:
[Red] Regions에 대한 테스트 추가

shows a selected button에 대한 테스트 추가.
: 서울(V)

TaeWoongKim2 committed:
[Refactor] index.jsx에 Provider 추가.

<Provider store={store}>
   <App />
</Provider>

TaeWoongKim2 committed:
[Refactor] Regions 컴포넌트에 대한 onClick 리팩토링

onClick={onClick(id)} -> onClick={() => onClick(id)} 로 변경.

Kihyun92 commented on src/RestaurantRegionContainer.jsx:
```diff
@@ -0,0 +1,31 @@
+
+import Regions from './Regions';
+
+export default function RestaurantRegionContainer() {
```
이 컴포넌트에서는 레스토랑에 대한 정보는 없고 region에 대한 정보만 있기 때문에 `RegionsContainer`라고만 해도 될 것 같은데 어떤가요??

TaeWoongKim2 replied:
레스토랑 앱이여서 계속 레스토랑을 붙이고 있었습니다..!! 지역에 대한 관심사만 있으니 `RegionsContainer` 좋네요!!

Kihyun92 commented on src/Regions.jsx:
```diff
@@ -0,0 +1,24 @@
+
+export default function Regions({ selectedRegionId, regions, onClick }) {
+  if (isEmpty(regions)) {
+    return <p>음식점 지역 정보를 등록해주세요!</p>;
```
정보가 없는 경우를 고려해보셨네요! 좋습니다 👍  지역 정보를 등록하는 기능은 아직 없기 때문에 `지역 정보가 없습니다` 같은 의미로 노출하는건 어떤가요?

TaeWoongKim2 replied:
간결하고 더 깔끔하네요!!

Kihyun92 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,72 @@
+    ));
+  }
+
+  context('with one or more regions', () => {
```
`with regions`라고만 하는건 어떤가요? :)

Kihyun92 commented on src/Regions.test.jsx:
```diff
@@ -0,0 +1,72 @@
+      { id: 3, name: '대구' },
+    ];
+
+    beforeEach(() => {
+      jest.clearAllMocks();
+    });
```
`context`보다 상위에서 관리하는게 좋을 것 같아요~

TaeWoongKim2 replied:
앗.. 미쳐 보지못한 미스가!! 감사합니다 :)

Kihyun92 commented on src/index.jsx:
```diff
@@ -1,9 +1,17 @@
+
 ReactDOM.render(
-  <App />,
+  (
+    <Provider store={store}>
+      <App />
+    </Provider>
+  ),
```
소괄호는 없어도 괜찮을 것 같아요~

TaeWoongKim2 replied:
소괄호가 없어도 되는군요..!!

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,29 @@
+}
+
+export default function reducer(state = initialState, action) {
+  if (typeof action === 'undefined') {
+    return state;
+  }
```
`unknownReducer` 함수로 아래에서 처리해줬기 때문에 이 부분은 필요 없지 않을까요?

TaeWoongKim2 replied:
삭제 후 조금 더 명료하게 변경했습니다 ^^

Kihyun92 reviewed (changes requested):
`Red-Green-Refactoring` 사이클을 잘 진행하고 계신것 같아요~ 나머지 기능들도 차근차근 추가해보시죠 😁

TaeWoongKim2 committed:
[Refactor] RegionsContainer 컴포넌트로 이름 변경

RegionsContainer로 컴포넌트 이름 변경
이에 따른 테스트 및 컴포넌트, App 수정.

TaeWoongKim2 committed:
[Refactor] Regions 컴포넌트에 대한 리팩토링

1st, jest.clearAllMocks(); describe() 바로 아래로 위치 변경.
2nd, Empty case에 대한 메시지 변경. (지역 정보가 없습니다).
*More simply, more clearly!

TaeWoongKim2 committed:
[Refactor] index.jsx 소괄호 제거

index.jsx 에서 Provider를 감싸는 불필요한 소괄호 제거.

TaeWoongKim2 committed:
[Refactor] unknownReducer 불필요한 함수 제거

불필요한 코드 제거

TaeWoongKim2 committed:
[Red] reducer setRegions에 대한 테스트 추가

fetch get /regions를 담을 setRegions에 대한 테스트 추가.

TaeWoongKim2 committed:
[Green] setRegions 구현

reducer에 setRegions 구현.

TaeWoongKim2 committed:
[Refactor] reducers parameter 리팩토링

const { type, payload } = action; 추가.
이에 따른 리팩토링 진행.

TaeWoongKim2 committed:
[Refactor] GET Regions API 구현

지역 정보를 가져오기 위해 GET fetch regions API 호출.

TaeWoongKim2 committed:
[Refactor] Redux-thunk 적용

미들웨어인 redux-thunk를 store.js에 적용.

TaeWoongKim2 committed:
[Refactor] loadRegions redux-thunk 분리

1st, App 컴포넌트에서 loadRegions({ dispatch }); 를 action.js로 분리.
2nd, loadRegions() redux-thunk 메소드 생성.

TaeWoongKim2 committed:
[Refactor] RegionsContainer 컴포넌트 제목 변경

Restaurant Regions -> Regions 로 변경에 따른 테스트 및 컴포넌트 수정.

TaeWoongKim2 committed:
[Red] App 컴포넌트 새로운 테스트 작성

초기 상태가 마운트되는지 테스트.

TaeWoongKim2 commented:
## 오늘 한 일

- TDD로 기능 구현

- 레스토랑 지역 목록 보기 기능 구현


## 해야할 일

- TDD로 기능 구현

-  레스토랑 카테고리 목록 보기 기능

-  레스토랑 목록 보기 기능

-  커버리지 100% 달성


## 오늘의 회고

> API를 받아오는 것을 연습해봤다. async / await 및 fetch 함수를 경험해볼 수 있는 값진 연습이었다.
API를 받아오는 것은 간단했으나, redux-thunk 라는 또 다시 새로운 기능으로 인해 쉽게 사용하기는 어려웠다.
조금 더 보기 편한 조금 더 명료한 코드를 짜기 위해 노력중이지만, 생각만큼 쉽지 않다.
이번 주만 두 번째 철야로 상당히 힘들지만, 금요일과 토요일 많은 시간을 투자해서 과제 및 개념정리를 할 생각이다.

## 질문사항

- 저의 경우, __mock__ 디렉토리를 홈 디렉토리(src의 부모 디렉토리)로 옮기는 순간 모든 test들이 오류가 발생합니다.
강의를 봤을 때, 옮기기만 해도 name space로 알아서 인식한다고 봤던 것 같은데 아닌가요??

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,25 @@
+};
+
+const reducers = {
+  setRegions: (state, payload) => ({
+    ...state,
+    regions: payload.regions,
+  }),
```
```suggestion
  setRegions: (state, { regions }) => ({
    ...state,
    regions,
  }),
```
이런식으로 구조 분해 할당을 사용해줘도 좋을 것 같아요~

TaeWoongKim2 replied:
오.. 요게 더 깔끔하네요 ㅎㅎ..

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,25 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (!action || !reducers[action.type]) {
```
`action`이 없는 경우에 대해서 처리해주셨네요! 이런 방법도 좋지만, 예외처리가 복잡해지는걸 막기 위해서 우리는 initialState를 추가한 것 처럼  default 값을 넣어주거나 혹은 하단에서 return 해줄때 예외처리를 해줄 수 있을 것 같아요~

한번 if 문을 사용하지 않고 표현할 수 있는지 고민해보시면 좋을 것 같습니다~

TaeWoongKim2 replied:
initialState를 추가한 것 처럼 default 값을 활용하여 변경했습니다 ^^
return 해줄때 예외 처리도 생각해보려고 했는데, `if (!reducers[action.type]) {}` 부분에서 action.type undefined 라고 오류가 발생하더라구요. 그래서 default parameter로 변경했습니다 : )

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,50 @@
+import { selectRegion, setRegions } from './actions';
+
+describe('reducer', () => {
+  context('with initial state', () => {
```
지금 같은 상황에서는 context로 분리하는게 아니라 describe로 분리하는게 좋을 것 같아요~ 

context는 주로 with <-> without

when log in <-> when log out

처럼 같은 맥락에서 별개의 사례일 때 context를 사용하거나, it 만으로는 설명이 너무 길어지는 경우에 사용하기도 합니다.

context 사용법은 아래 문서와 아샬님의 문서를 참고해주세요~

- https://www.betterspecs.org/
- https://github.com/ahastudio/til/blob/main/blog/2018/12-08-given-when-then.md

TaeWoongKim2 replied:
음 그렇군요. 참고해주신 문서도 읽어보겠습니다!

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,50 @@
+
+      const state = reducer(initialState, setRegions(regions));
+
+      expect(state.regions).toHaveLength(7);
```
region 상태가 바뀌는 경우에 대해 테스트를 해주셨네요! 

이때 7이라는 고정된 값보다는 상태가 바뀔때마다 대응이 가능하도록 처리할 순 없을까요?

TaeWoongKim2 replied:
`regions.length` 로 상태가 바뀔때마다 대응이 가능하도록  변경했습니다!

Kihyun92 commented on src/services/api.js:
```diff
@@ -0,0 +1,16 @@
+  const url = 'https://eatgo-customer-api.ahastudio.com/regions';
+  const response = await fetch(url);
+  const data = await response.json();
+  console.log(data);
```
커밋시 불필요한 콘솔 로그는 지워주세요~

Kihyun92 reviewed (changes requested):
답변이 늦어 죄송합니다 ㅠㅠ 바쁜 한 주를 보내고 계신 것 같네요! 짬내서 과제를 하시는게 멋지네요 👍

전반적으로 `Red-Green-Refactoring` 사이클을 잘 지키려 노력하신 것 같아요 :)
주말이 얼마 남지 않았지만, 남은 기능들도 차근차근 진행해보면 좋을 것 같아요~

__mocks__ 폴더를 프로젝트 최상단에 위치시킨다는 말씀같아요~ 혹시 테스트코드를 종료하고 다시 실행해보셨나요? 계속 발생한다면 터미널에 어떻게 에러가 발생하는지 로그를 보여주시면 더 좋을 것 같아요 ☺️

TaeWoongKim2 committed:
[Refactor] Code Review에 대한 코드 리팩토링

더 간결하고 더 명시적인 코드를 위한 리팩토링

TaeWoongKim2 committed:
[Red] CategoriesContainer에 대한 테스트 작성

1st, CategoriesContainer.test.jsx 생성.
2st, __fixture__/categories 임시 테스트 데이터 생성.
3st, CategoriesContainer.jsx 컴포넌트 파일 생성.

TaeWoongKim2 committed:
[Green] CategoriesContainer에 구현

TDD 기반 CategoriesContainer 컴포넌트 구현.
*TODO: Categories 테스트 작성 및 구현.

TaeWoongKim2 committed:
[Red] Categories에 대한 테스트 작성

Categories 컴포넌트에 대한 테스트 작성.
*TODO: Make Categories component!

TaeWoongKim2 committed:
[Green] Categories 컴포넌트 구현

TDD 기반의 Categories 컴포넌트 구현.
1st, show categories of food.
2nd, click category button.
3rd, shows a selected button.
4th, show nothing message without categories.

*TODO: Apply CategoriesContainer into App.

TaeWoongKim2 committed:
[Green] App에 CategoriesContainer 반영

*TODO: 카테고리 클릭시, 체크하는 reducer 생성.
*RED -> Green

TaeWoongKim2 committed:
[Red] reducer, CategoriesContainer에 대한 action 테스트 작성

1st, if select a category.
2nd, changes categories.

*TODO: Make action creator and apply reducers into App.

TaeWoongKim2 committed:
[Green] CategoriesContainer에 대한 reducers 구현

Make action creator: setCategories & selectCategory.

TaeWoongKim2 committed:
[Refactor] CategoriesContainer에 react-redux 리팩토링

useDispatch, useSelector 적용.
1st, useSelector for get initial states named 'selectedCategoryId' and 'categories'.
2nd, useDispatch for get action creator named 'selectCategory'.

*TODO: fetch Categories API.

TaeWoongKim2 committed:
[Refactor] Categories API 를 위한 action creator 구현

1st, Make action creator named 'loadCategories'.
2nd, Make fetch api named 'fetchCategories'.

*TODO: Apply into App.

TaeWoongKim2 committed:
[Refactor] App에 fetch Categories 반영 및 오류제거

1st, dispatch(loadCategories()); 적용, fetch categories onload.
2nd, action.js 리팩토링. replaceAll 기능 사용으로 인한 오류 수정.
3nd, App.test.js 테스트 리팩토링, expect(dispatch).toBeCalledTimes(2);, 1-> 2

*TODO: Make restaurantsContainer component and fetch restaurants!

TaeWoongKim2 committed:
[Refector] API fetch 테스트 코드에 대한 고민

TaeWoongKim2 commented:
## 오늘 한 일

- 코드 리뷰 해주신 부분 반영!

- TDD로 기능 구현!

-  레스토랑 카테고리 목록 보기 기능!


## 오늘 해야할 일

- TDD로 기능 구현

-  레스토랑 목록 보기 기능

-  커버리지 100% 달성


## 고민중인 것

- API에 대한 테스트는 어떻게 진행할 것인가? 
> 이 부분에 대해 좀 생각해보고 찾아서 커버리지 100%를 달성할 예정이다.

TaeWoongKim2 committed:
[Refector] App에 대한 테스트 수정

목업 초기화 코드 추가.
beforeEach(() => jest.clearAllMocks());

TaeWoongKim2 committed:
[Red] RestaurantsContainer에 대한 테스트 작성

1st, shows a title.
*TODO: Check API fetch correctly.

TaeWoongKim2 committed:
[Green] RestaurantsContainer 구현

* TODO: Make Restaurants test and component!

TaeWoongKim2 committed:
[Red] Restaurants에 대한 테스트 작성

1st, make test.
   - case.1  with restaurants
   - case.2  without restaurants
2st, make restaurants initial data in __fixture__.

*TODO:  make restaurants component!

TaeWoongKim2 committed:
[Green] Restaurants 컴포넌트 구현

Restaurants 테스트를 기반으로 Restaurants 컴포넌트 구현.
* TODO: apply redux in RestaurantsContainer!

TaeWoongKim2 committed:
[Refactor] RestaurantsContainer 리팩토링.

RestaurantsContainer 리팩토링 및 Restaurants에 대한 테스트 수정.

TaeWoongKim2 committed:
[Refactor] RestaurantsContainer 컴포넌트 리팩토링.

레스토랑 목록 API를 받기 위한 사전 리팩토링.
*TODO: fetch /restaurants?region=${regionName}&category=${categoryId}

TaeWoongKim2 committed:
[Red & Green] RestaurantsContainer에 대한 테스트 추가 및 구현.

1st, 지역과 분류 모두 선택이 안된 경우, '지역과 분류를 선택해주세요' 메시지 출력 구현.
2nd, 이에 대한 테스트 작성.

*TODO: API

TaeWoongKim2 committed:
[Refactor] RestaurantsContainer에 loadRestaurants 구현.

*TODO actions에 해당 API fetch 구현.

TaeWoongKim2 committed:
[Refactor] fetchRestaurants api function 구현

1st, Make action creator name 'loadRestaurants()' and 'setRestaurants()'
2nd, Make API function named 'fetchRestaurants()'

TaeWoongKim2 committed:
[Refactor]

TaeWoongKim2 committed:
[Refactor] 레스토랑 API를 위한 하위 컴포넌트 테스트 수정

TaeWoongKim2 committed:
[Red] setRestaurants에 대한 테스트 신규 테스트 작성

Coverage 100%를 위한 신규 테스트 작성.
*TODO: API Test 작성!

TaeWoongKim2 committed:
[Red] action에 대한 테스트 신규 추가

1st, loadRegions에 대한 테스트 작성.
2nd, loadCategories에 대한 테스트 작성.
3rd, loadRestaurants에 대한 테스트 작성.

*TODO: api에 대한 테스트 작성!

TaeWoongKim2 committed:
[Red] api에 대한 테스트 작성

Coverage 100% 달성을 위한 api.js 테스트 작성.
import { enableFetchMocks } from 'jest-fetch-mock'; 를 통해 API reponse mock 작업 완료.

TaeWoongKim2 commented:
## 오늘 한 일 👍 

- 코드 리뷰 해주신 부분 반영!

- TDD로 기능 구현!

-  레스토랑 카테고리 목록 보기 기능!

-  레스토랑 목록 보기 기능!

-  커버리지 100% 달성!


## 새롭게 발견한 것

> API load 및 fetch에 대한 테스트에 대해서 이해가 어려웠으며, 사용하기 상당히 어려웠다.

- `'redux-mock-store'`에 대한 것.

- `'jest-fetch-mock'`에 대한 것.


## 질문

- 로컬에서는 모든 테스트가 통과했으나, github에 올라가면, test 시에 All checks have failed가 떨어진다...
원인을 정확히 모르겠다 ㅠㅠ..

--> 컨테이너별 헤더를 지워야했다.... conceptJS에서 뭔가 순차적으로 검사하는 것 같다..

Kihyun92 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,68 @@
+
+  context('with categories', () => {
+    it('show categories of food', () => {
+      const { getByText } = renderCategories(0, intialCategories);
```
renderCategories에 0을 바로 넣어주는 것도 좋지만, 
상위에 `const selectedCategoryId = 0`을 선언해서 어떤 값을 넣어주는건지 명확하게 의도를 나타내줘도 좋을 것 같네요 :)

TaeWoongKim2 replied:
더 의도를 명확하게 하기 위한 리팩토링은 언제나 환영이죠 !! 변경하겠습니다 :)

TaeWoongKim2 committed:
[Refactor] Remove Header

Github 테스트 통과를 위한 각 컨테이너별 제목 제거!

Kihyun92 commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  return (
+    <div>
```
div 태그로 감싸줄 필요 없을 것 같아요~

Kihyun92 commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  return (
+    <div>
```
여기도 제거해도 될 것 같네요~

Kihyun92 commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,35 @@
+  }));
+
+  describe('selects a region', () => {
+    it('checks selected region', () => {
```
describe와 it함수 내의 설명이 중복 되는 것 같습니다. 조금 더 이 테스트의 의도를 나타내면 좋을 것 같아요~
예를 들면 클릭하면 dispatch가 call 된다거나 하는 것 처럼요 :)

TaeWoongKim2 replied:
흠...  중복되는 말이긴 하네요!! call dispatch로 의도를 더욱 드러내겠습니다!

Kihyun92 commented on src/RestaurantsContainer.jsx:
```diff
@@ -0,0 +1,24 @@
+    restaurants: state.restaurants,
+  }));
+
+  if (!selectedRegionId || !selectedCategoryId) {
+    return <p>지역과 분류를 선택해주세요!</p>;
+  }
+
+  return (
+    <div>
+      <h4>레스토랑 목록</h4>
+      <Restaurants restaurants={restaurants} />
+    </div>
+  );
```
category와 region이 선택되지 않은 경우에 대한 처리를 해주셨네요 👍 

다만 부정문을 사용하는것보다 가능하면 긍정문으로(not을 사용하지 않는) 처리해보면 좋을 것 같아요~

```suggestion
  return (
    {  selectedRegionId && selectedCategoryId ? (
      <>
        <h4>레스토랑 목록</h4>
        <Restaurants restaurants={restaurants} />
      </>
    : (
      <p>지역과 분류를 선택해주세요!</p>
    )
  );
```

TaeWoongKim2 replied:
부정문을 사용하는것보다 가능하면 긍정문으로(not을 사용하지 않는) 처리하는게 가독성면에서도 좋은 것 같아요!!

Kihyun92 commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,34 @@
+describe('RestaurantsContainer', () => {
+  describe('render title', () => {
+    context('without selections', () => {
+      it('shows a message', () => {
```
어떤 메시지인지 명확하게 해주는건 어떨까요?

Kihyun92 reviewed (changes requested):
커버리지 100%를 만들기 위해 많은 고민을 해보신 것 같아요~! 👍 
API에 대한 테스트의 경우 어려운 점이 많았을 텐데 잘 해주신 것 같아요:)
API 테스트에 관해서는 이미 읽어보셨을 수 있지만, 아래 문서를 읽어보시면 좋을 것 같아요~
- https://jestjs.io/docs/en/mock-functions#mocking-modules

지금도 충분히 잘 해주셨지만 조금 더 다듬고 마무리하면 좋을 것 같습니다~

Kihyun92 commented:
과제는 잘 진행해보셨나요? 남은 부분은 스스로 진행해보시면 좋고, 아샬님의 강의를 보고 보완해보셔도 좋을 것 같습니다~

TaeWoongKim2 commented:
이번에도 많은 것을 익히고 배우는 시간을 가져서 좋네요.
다만, 지난 주부터 직장 업무가 차곡차곡 쌓여서 버거워지고 있네요 ㅎㅎ..!!
다음주도 힘내서 파이팅하겠습니다. 코드 리뷰 감사드립니다 :)

TaeWoongKim2 committed:
[Refactor] Apply the better code from Kihyun's code review.

More clearly and more simple!

- 긍정문으로 처리하는 것이 가독성이 더 좋다!
- 테스트는 항상 명시적으로 표시하자!
- 불필요한 부분은 항상 빠르게 제거하자!

Kihyun92 reviewed (approved):
바쁜 한주를 보내신 것 같아요~ 고생하셨습니다~ 쉽지 않은 과제였는데, 잘해주셨어요 👍
`Red-Green-Refactoring` 사이클도 잘 적용하시고, Redux와 API 연동 부분을 잘 수행해주신 것 같습니다~ 과제 풀이 영상도 꼭 보시면 좋겠습니다.
이번 과제를 통해 어떤걸 배우셨는지, 그리고 더 배워보고 싶은 부분이 있나요? 😁

TaeWoongKim2 commented:
## 한 일(Facts)
1. TDD Cycle 반복 연습
    - 어플리케이션 구현을 TDD 방식으로 반복 연습을 했다!
    - 비동기 처리 방식에 대한 이해
2. async / await 에 대한 활용법 숙지했다!
    - fetch API 활용했다!
3. 과제 5. 레스토랑 조회 구현
    - TDD 방식으로 관심사를 나눠 차근차근 구현해보았다!
    - API 비동기 테스트에 대한 새로운 개념 숙지 및 활용했다!
    - Coverage 100% 달성을 위한 JEST 공식문서 리딩했다!

## 배운 점 (Findings)
    - async / await, fetch 를 통한 비동기 처리
    비동기 처리는 항상 jQuery를 통해서 처리했었는데, 모던 자바스크립트의 fetch API를 숙지하고 간편하게 사용해봄으로써 순수 자바스크립트만으로도 비동기 처리함에 있어 어렵지 않음을 경험했다!

    - 비동기 처리에 대한 mocking 테스트 방법
    `mockResolvedValue` 를 통해 가짜 응답을 만듬으로써 테스트를 할 수 있었다!

    - 비동기 처리를 위한 미들웨어 redux thunk 사용 방법
    미들웨어를 통해 함수를 디스패치할 수 있고, 디스패치된 함수에 dispatch와 getState를 파라미터로 받아와 비동기 처리에 활용할 수 있었다!


## 배우고 싶은 점
배우고 싶은 점이라기 보다는 복습하고 싶습니다.
async / await 에 대해서 더 공부해보고 싶고, Promise와 then도 사용해본 적이 없어서 이 문법은 어떠한 것인지도 파악해보고 싶네요!!
모던 자바스크립트에 대해 더욱 깊게 이해해보고 싶어졌습니다 :)

Kihyun92 commented:
이번 과제를 하시면서 많은걸 느끼신 것 같네요 ㅎㅎ 꼭 복습해보셨으면 좋겠습니다~ 화이팅 👍
